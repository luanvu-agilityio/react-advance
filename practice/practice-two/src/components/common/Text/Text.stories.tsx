import type { Meta, StoryObj } from '@storybook/react'
import Text from './index'

const meta = {
  title: 'Components/Common/Text',
  component: Text,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    text: { control: 'text' },
    as: {
      control: 'select',
      options: ['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'div'],
    },
    className: { control: 'text' },
    style: { control: 'object' },
  },
} satisfies Meta<typeof Text>

export default meta
type Story = StoryObj<typeof Text>

// Basic paragraph
export const Paragraph: Story = {
  args: {
    text: 'This is a paragraph text',
    as: 'p',
  },
}

// Heading 1
export const Heading1: Story = {
  args: {
    text: 'This is a Heading 1',
    as: 'h1',
    style: {
      fontSize: '2rem',
      fontWeight: 'bold',
    },
  },
}

// Heading 2
export const Heading2: Story = {
  args: {
    text: 'This is a Heading 2',
    as: 'h2',
    style: {
      fontSize: '1.5rem',
      fontWeight: 'bold',
    },
  },
}

// Span text
export const SpanText: Story = {
  args: {
    text: 'This is an inline span text',
    as: 'span',
  },
}

// Styled text
export const StyledText: Story = {
  args: {
    text: 'This is a styled text',
    as: 'p',
    style: {
      color: '#2563eb',
      fontSize: '1.25rem',
      fontWeight: '600',
      textDecoration: 'underline',
    },
  },
}

// Multiple text variants
export const TextVariants: Story = {
  render: () => (
    <div style={{ display: 'flex', flexDirection: 'column', gap: '1rem' }}>
      <Text
        as="h1"
        text="Heading 1"
        style={{ fontSize: '2rem', fontWeight: 'bold' }}
      />
      <Text
        as="h2"
        text="Heading 2"
        style={{ fontSize: '1.5rem', fontWeight: 'bold' }}
      />
      <Text
        as="h3"
        text="Heading 3"
        style={{ fontSize: '1.25rem', fontWeight: 'bold' }}
      />
      <Text as="p" text="Regular paragraph text" />
      <Text as="span" text="Inline span text" />
      <Text
        as="div"
        text="Custom styled div"
        style={{
          backgroundColor: 'var(--coral-shade-3)',
          padding: '0.5rem',
          borderRadius: '0.25rem',
        }}
      />
    </div>
  ),
}
