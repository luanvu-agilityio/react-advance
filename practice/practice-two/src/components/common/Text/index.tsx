import type { HTMLAttributes, CSSProperties } from 'react'

interface TextProps extends HTMLAttributes<HTMLElement> {
  text: string
  as?: 'p' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'div'
  className?: string
  style?: CSSProperties
}

const Text = ({
  text,
  as = 'p',
  className,
  style,
  children,
  ...rest
}: TextProps) => {
  const Element = as

  return (
    <Element className={className} style={style} {...rest}>
      {text ?? children}
    </Element>
  )
}
export default Text
