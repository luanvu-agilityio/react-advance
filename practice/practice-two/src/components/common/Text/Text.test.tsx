import { render } from '@testing-library/react'
import Text from './index'

describe('Text Component', () => {
  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders basic text variants correctly', () => {
      // Default paragraph rendering
      const { container: defaultParagraph } = render(
        <Text text="Default paragraph" />
      )
      expect(defaultParagraph).toMatchSnapshot('default-paragraph')

      // Rendering with children instead of text prop
      const { container: withChildren } = render(
        <Text text="">Child content</Text>
      )
      expect(withChildren).toMatchSnapshot('with-children')

      // Custom class name
      const { container: withClass } = render(
        <Text text="With class" className="custom-class" />
      )
      expect(withClass).toMatchSnapshot('with-class')
    })

    it('renders with different HTML elements', () => {
      // Different element types
      const elements = [
        'h1',
        'h2',
        'h3',
        'h4',
        'h5',
        'h6',
        'span',
        'div',
      ] as const

      elements.forEach((element) => {
        const { container } = render(
          <Text text={`${element} text`} as={element} />
        )
        expect(container).toMatchSnapshot(`as-${element}`)
      })
    })

    it('renders with custom styles and attributes', () => {
      // With custom styles
      const customStyle = { color: 'red', fontSize: '16px' }
      const { container: styledContainer } = render(
        <Text text="Styled text" style={customStyle} />
      )
      expect(styledContainer).toMatchSnapshot('with-custom-style')

      // With additional attributes
      const { container: withAttributes } = render(
        <Text
          text="With attributes"
          data-testid="custom-text"
          aria-label="text label"
        />
      )
      expect(withAttributes).toMatchSnapshot('with-attributes')

      // Long text content
      const longText =
        'This is a very long text that should test how the component handles wrapping and display of lengthy content when it potentially spans multiple lines.'
      const { container: longTextContainer } = render(<Text text={longText} />)
      expect(longTextContainer).toMatchSnapshot('with-long-text')
    })
  })
})
