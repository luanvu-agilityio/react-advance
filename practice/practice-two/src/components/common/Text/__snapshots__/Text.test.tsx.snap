// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Rendering renders basic text variants correctly: default-paragraph 1`] = `
<div>
  <p>
    Default paragraph
  </p>
</div>
`;

exports[`Text Component Rendering renders basic text variants correctly: with-children 1`] = `
<div>
  <p />
</div>
`;

exports[`Text Component Rendering renders basic text variants correctly: with-class 1`] = `
<div>
  <p
    class="custom-class"
  >
    With class
  </p>
</div>
`;

exports[`Text Component Rendering renders with custom styles and attributes: with-attributes 1`] = `
<div>
  <p
    aria-label="text label"
    data-testid="custom-text"
  >
    With attributes
  </p>
</div>
`;

exports[`Text Component Rendering renders with custom styles and attributes: with-custom-style 1`] = `
<div>
  <p
    style="color: red; font-size: 16px;"
  >
    Styled text
  </p>
</div>
`;

exports[`Text Component Rendering renders with custom styles and attributes: with-long-text 1`] = `
<div>
  <p>
    This is a very long text that should test how the component handles wrapping and display of lengthy content when it potentially spans multiple lines.
  </p>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-div 1`] = `
<div>
  <div>
    div text
  </div>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h1 1`] = `
<div>
  <h1>
    h1 text
  </h1>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h2 1`] = `
<div>
  <h2>
    h2 text
  </h2>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h3 1`] = `
<div>
  <h3>
    h3 text
  </h3>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h4 1`] = `
<div>
  <h4>
    h4 text
  </h4>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h5 1`] = `
<div>
  <h5>
    h5 text
  </h5>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-h6 1`] = `
<div>
  <h6>
    h6 text
  </h6>
</div>
`;

exports[`Text Component Rendering renders with different HTML elements: as-span 1`] = `
<div>
  <span>
    span text
  </span>
</div>
`;
